From f61dec14151d9fc864e91b9b430ac0c2576a9d5e Mon Sep 17 00:00:00 2001
From: Chris Burroughs <chris.burroughs@joyent.com>
Date: Fri, 4 Nov 2016 11:52:12 -0400
Subject: [PATCH] joyent/node-manta#272 proudly proclaim the latest version
 joyent/node-manta#282 rely on dashdash for mchmod arg parsing

---
 bin/mchattr          |  2 ++
 bin/mchmod           | 21 +++++++++++----------
 bin/mfind            |  2 ++
 bin/mget             |  2 ++
 bin/minfo            |  2 ++
 bin/mjob             | 21 +++++++++++++++++++--
 bin/mln              |  2 ++
 bin/mlogin           |  2 ++
 bin/mls              |  2 ++
 bin/mmd5             |  2 ++
 bin/mmkdir           |  2 ++
 bin/mput             |  2 ++
 bin/mrm              |  2 ++
 bin/mrmdir           |  2 ++
 bin/msign            |  2 ++
 bin/muntar           |  2 ++
 lib/create_client.js | 15 ++++++++++++++-
 lib/index.js         |  1 +
 18 files changed, 73 insertions(+), 13 deletions(-)

diff --git a/bin/mchattr b/bin/mchattr
index 57a8e60..6072edc 100755
--- a/bin/mchattr
+++ b/bin/mchattr
@@ -60,6 +60,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mchmod b/bin/mchmod
index aadd417..bf87a7f 100755
--- a/bin/mchmod
+++ b/bin/mchmod
@@ -40,20 +40,20 @@ function ifError(err) {
 }
 
 
-function parseMchmodOptions() {
-    if (process.argv.length <= 2) {
+function parseMchmodOptions(args) {
+    if (args.length < 2) {
         manta.cli_usage(OPTIONS_PARSER, 'role required',
                                         '-- [+-=]role,... path...');
     }
 
-    var startRoles = process.argv.indexOf('--');
+    var startRoles = args.indexOf('--');
     var operation;
     var roleOpt;
 
     if (startRoles < 0) {   // no other options: `mchmod +role path`
-        roleOpt = process.argv.splice(2, 1).pop();
+        roleOpt = args.splice(0, 1).pop();
     } else {                // other options: `mchmod -a account -- +role path`
-        roleOpt = process.argv.splice(startRoles, 2).pop();
+        roleOpt = args.splice(startRoles, 2).pop();
     }
 
     operation = roleOpt.charAt(0);
@@ -72,8 +72,6 @@ function parseMchmodOptions() {
 function parseOptions() {
     var opts;
 
-    var mchmodOpts = parseMchmodOptions();
-
     try {
         opts = OPTIONS_PARSER.parse(process.argv);
         manta.checkBinEnv(opts);
@@ -82,9 +80,6 @@ function parseOptions() {
                                         '-- [+-=]role,... path...');
     }
 
-    opts.mchmodOperation = mchmodOpts.operation;
-    opts.mchmodRoles = mchmodOpts.roles;
-
     manta.cli_logger(opts, LOG);
 
     if (opts.help) {
@@ -92,6 +87,12 @@ function parseOptions() {
                                         '-- [+-=]role,... path...');
     }
 
+    manta.handleOptsVersion(opts);
+
+    var mchmodOpts = parseMchmodOptions(opts._args);
+    opts.mchmodOperation = mchmodOpts.operation;
+    opts.mchmodRoles = mchmodOpts.roles;
+
     if (opts._args.length < 1) {
         manta.cli_usage(OPTIONS_PARSER, 'path required',
                                         '-- [+-=]role,... path...');
diff --git a/bin/mfind b/bin/mfind
index c213f2a..470cd4b 100755
--- a/bin/mfind
+++ b/bin/mfind
@@ -109,6 +109,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, '[PATH]');
 
+    manta.handleOptsVersion(opts);
+
     if (opts.name)
         opts.name = new RegExp(opts.name);
 
diff --git a/bin/mget b/bin/mget
index 2015526..fef3555 100755
--- a/bin/mget
+++ b/bin/mget
@@ -82,6 +82,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/minfo b/bin/minfo
index 3876c37..d865b18 100755
--- a/bin/minfo
+++ b/bin/minfo
@@ -53,6 +53,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mjob b/bin/mjob
index e252c56..b8c8499 100755
--- a/bin/mjob
+++ b/bin/mjob
@@ -46,6 +46,8 @@ function assertOpts(thisp, subcmd, opts, cb) {
         return (false);
     }
 
+    manta.handleOptsVersion(opts);
+
     try {
         manta.checkBinEnv(opts);
         manta.cli_logger(opts, LOG);
@@ -429,11 +431,26 @@ function getJobInfo(opts, cb) {
 function MJob() {
     cmdln.Cmdln.call(this, {
         name: 'mjob',
-        desc: 'Manages Manta compute jobs'
-    });
+        desc: 'Manages Manta compute jobs',
+        options: [
+            {
+                names: ['help', 'h'],
+                type: 'bool',
+                help: 'Show this help message and exit.'
+            },
+            {
+                names: ['version'],
+                type: 'bool',
+                help: 'Print version and exit.'
+            }
+        ]});
 }
 util.inherits(MJob, cmdln.Cmdln);
 
+MJob.prototype.init = function (opts, args, callback) {
+    manta.handleOptsVersion(opts);
+    cmdln.Cmdln.prototype.init.apply(this, arguments);
+};
 
 //-- AddKeys
 
diff --git a/bin/mln b/bin/mln
index 1dc3164..df4ff60 100755
--- a/bin/mln
+++ b/bin/mln
@@ -61,6 +61,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'source dest');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'source required', 'source dest');
     if (opts._args.length < 2)
diff --git a/bin/mlogin b/bin/mlogin
index e131a77..38e3835 100755
--- a/bin/mlogin
+++ b/bin/mlogin
@@ -124,6 +124,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length > 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mls b/bin/mls
index f9ed516..c97c088 100755
--- a/bin/mls
+++ b/bin/mls
@@ -94,6 +94,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, '[PATH]...');
 
+    manta.handleOptsVersion(opts);
+
     opts.paths = [];
     if (opts._args.length < 1) {
         opts.paths = ['/' + opts.account + '/stor'];
diff --git a/bin/mmd5 b/bin/mmd5
index 165559a..b70a10d 100755
--- a/bin/mmd5
+++ b/bin/mmd5
@@ -56,6 +56,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mmkdir b/bin/mmkdir
index 10dd55a..396c599 100755
--- a/bin/mmkdir
+++ b/bin/mmkdir
@@ -71,6 +71,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     opts.headers = {};
     (opts.header || []).forEach(function (h) {
         if (h.indexOf(':') === -1) {
diff --git a/bin/mput b/bin/mput
index 991344b..d177572 100755
--- a/bin/mput
+++ b/bin/mput
@@ -113,6 +113,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts.md5 && !opts.file)
         manta.cli_usage(OPTIONS_PARSER, false, '--md5 requires --file');
 
diff --git a/bin/mrm b/bin/mrm
index 9925fa1..73be313 100755
--- a/bin/mrm
+++ b/bin/mrm
@@ -66,6 +66,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mrmdir b/bin/mrmdir
index f3e7768..a3401f2 100755
--- a/bin/mrmdir
+++ b/bin/mrmdir
@@ -51,6 +51,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/msign b/bin/msign
index 6e3d490..9d6f493 100755
--- a/bin/msign
+++ b/bin/msign
@@ -89,6 +89,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.handleOptsVersion(opts);
+
     opts.method = opts.method.map(function (m) {
         m = m.toUpperCase();
         if (METHODS.indexOf(m) === -1) {
diff --git a/bin/muntar b/bin/muntar
index 41d6060..63c97e7 100755
--- a/bin/muntar
+++ b/bin/muntar
@@ -100,6 +100,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path');
 
+    manta.handleOptsVersion(opts);
+
     if (!opts.file)
         manta.cli_usage(OPTIONS_PARSER, 'file is a required argument', 'path');
 
diff --git a/lib/create_client.js b/lib/create_client.js
index 86b753b..c0c714a 100644
--- a/lib/create_client.js
+++ b/lib/create_client.js
@@ -11,6 +11,7 @@ var restifyClients = require('restify-clients');
 var auth = require('smartdc-auth');
 var manta = require('./client');
 
+var packageJson = require('../package.json');
 
 
 ///--- Globals
@@ -42,6 +43,11 @@ var DEFAULT_OPTIONS = [
         type: 'bool',
         help: 'Print this help and exit'
     },
+    {
+        name: 'version',
+        type: 'bool',
+        help: 'Print version and exit.'
+    },
     {
         names: ['insecure', 'i'],
         type: 'bool',
@@ -272,6 +278,12 @@ function setupLogger(opts, log) {
 }
 
 
+function handleOptsVersion(opts) {
+    if (opts.version) {
+        console.log(packageJson.version);
+        process.exit(0);
+    }
+}
 
 ///--- Exports
 
@@ -283,7 +295,8 @@ module.exports = {
     createBinClient: createBinClient,
     DEFAULT_OPTIONS: DEFAULT_OPTIONS,
     usage: usage,
-    setupLogger: setupLogger
+    setupLogger: setupLogger,
+    handleOptsVersion: handleOptsVersion
 };
 
 
diff --git a/lib/index.js b/lib/index.js
index da64fb1..6b9ed83 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -43,6 +43,7 @@ module.exports = {
     DEFAULT_CLI_OPTIONS: cc.DEFAULT_OPTIONS,
     cli_usage: cc.usage,
     cli_logger: cc.setupLogger,
+    handleOptsVersion: cc.handleOptsVersion,
     StringStream: StringStream,
     path: manta.path,
     jobPath: manta.jobPath,
-- 
2.21.0

